packages/terra-image/src/Image.jsx

```diff
@@ -47,36 +51,106 @@ const defaultProps = {
   alt: ' ',
 };
 
-const Image = ({
-  src,
-  variant,
-  isFluid,
-  alt,
-  height,
-  width,
-  onLoad,
-  onError,
-  ...customProps
-}) => {
-  const classes = cx([
-    'image',
-    variant,
-    { fluid: isFluid },
-    customProps.className,
-  ]);
-
-  return (
-    <img
-      src={src}
-      alt={alt}
-      height={height}
-      width={width}
-      onLoad={onLoad}
-      onError={onError}
-      {...customProps}
-      className={classes}
-    />);
-};
+class Image extends React.Component {
+
+  constructor(props) {
+    super(props);
+
+    this.state = { isLoading: true, isError: false };
+    const { src, variant, isFluid, alt, placeholder, height, width, onLoad, onError, ...customProps } = this.props;
+
+    this.classes = cx([
+      'image',
+      variant,
+      customProps.className,
+      { fluid: isFluid },
+    ]);
+
+    this.handleOnLoad = this.handleOnLoad.bind(this);
+    this.handleOnError = this.handleOnError.bind(this);
+  }
+
+  componentWillReceiveProps(newProps) {
+    // If a new image is being loaded, reset the state to loading
+    if (newProps.src !== this.props.src) {
+      this.setState({ isLoading: true, isError: false });
+    }
+  }
+
+  handleOnLoad() {
+    this.setState({ isLoading: false });
+    const onLoad = this.props.onLoad;
+    if (onLoad !== undefined) {
+      onLoad();
+    }
+  }
+
+  handleOnError() {
+    this.setState({ isLoading: false, isError: true });
+    const onError = this.props.onError;
+    if (onError !== undefined) {
+      onError();
+    }
+  }
+
+  createPlaceholderImage(customProps) {
+    const { placeholder, alt, height, width } = this.props;
+    return (
+      <img
+        src={placeholder}
+        alt={alt}
+        height={height}
+        width={width}
+        className={this.classes}
+        {...customProps}
+      />
+    );
+  }
+
+  createImage(customProps) {
+    const { src, alt, height, width } = this.props;
+    return (
+      <img
+        src={src}
+        alt={alt}
+        height={height}
+        width={width}
+        onLoad={this.handleOnLoad}
+        onError={this.handleOnError}
+        className={this.classes}
+        {...customProps}
+      />
+    );
+  }
+
+  render() {
+    const { src, variant, isFluid, alt, placeholder, height, width, onLoad, onError, ...customProps } = this.props;
+    delete customProps.className;
+
+    if (placeholder) {
+      if (this.state.isLoading) {
+        return (
+          <div>
+            <div className={styles.hidden}>{this.createImage(customProps)}</div>
+            <div>{this.createPlaceholderImage(customProps)}</div>
+          </div>
+        );
+      }
+
+      return (
+        <div>
+          <div>{this.state.isError ? this.createPlaceholderImage(customProps) : this.createImage(customProps)}</div>
```

Comment:
What if it's the placeholderImage that threw an error here, and not the original image?

Reply:
as of now, the placeholder rendering cannot trigger isError to be set to true I don't think. maybe that is a logic flow that needs to be explored more, with the discussion of a fallback for the placeholder

Reply:
This may not have been fully established, and since we don't have a universal fallback atm, I would be okay if we log an issue to come back to it.